### PR TITLE
SAA-422 expiring auto-suspended allocations when they are 5 days or more old.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
A separate PR is needed to add a new attribute to the prison regime for the max number of days expiry window.

It is currently hard-coded to 5 days.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/PrisonRegime.kt [hunk 1]
```diff
@@ -27,5 +29,11 @@ data class PrisonRegime(
   val edStart: LocalTime,
 
   val edFinish: LocalTime,
-
-)
+) {
+  // TODO the expiry days needs to be configurable here i.e. new column is required to capture this.
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/PrisonRegime.kt:33]: To be addressed in a separate PR.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/BulkAppointmentIntegrationTest.kt [hunk 2]
```diff
@@ -43,21 +43,21 @@ class BulkAppointmentIntegrationTest : IntegrationTestBase() {
...
     prisonApiMockServer.stubGetAppointmentScheduleReasons()
-    prisonApiMockServer.stubGetLocationsForAppointments(request.prisonCode!!, request.internalLocationId!!)
+    prisonApiMockServer.stubGetLocationsForAppointments(request.prisonCode, request.internalLocationId)
     prisonerSearchApiMockServer.stubSearchByPrisonerNumbers(
       prisonerNumbers,
       listOf(
         PrisonerSearchPrisonerFixture.instance(
           prisonerNumber = prisonerNumbers[0],
           bookingId = 1,
-          prisonId = request.prisonCode!!,
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/BulkAppointmentIntegrationTest.kt:55]: Removing warnings.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReleasedEventHandler.kt [hunk 3]
```diff
@@ -28,24 +28,27 @@ class OffenderReleasedEventHandler(
...
           log.warn("Failed to handle event $event")
+          false
         }
       }
     } else {
       log.info("Ignoring released event for ${event.prisonCode()} - not rolled out.")
     }
 
-    return false
+    return true
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReleasedEventHandler.kt:51]: Should be returning true if prison is not rolled.  There is no further action to be taken with the event in this situation.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReceivedEventHandler.kt [hunk 4]
```diff
@@ -40,11 +40,11 @@ class OffenderReceivedEventHandler(
...
         return true
       }
-    } else {
-      log.info("Ignoring received event for ${event.prisonCode()} - not rolled out.")
     }
 
-    return false
+    log.info("Ignoring received event for ${event.prisonCode()} - not rolled out.")
+
+    return true
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReceivedEventHandler.kt:47]: Should be returning true if prison is not rolled.  There is no further action to be taken with the event in this situation.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonersearchapi/api/PrisonerSearchApiApplicationClient.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.client.prisonersearchapi.api
+
+import org.springframework.stereotype.Service
+import org.springframework.web.reactive.function.client.WebClient
+
+@Service
+class PrisonerSearchApiApplicationClient(prisonerSearchApiAppWebClient: WebClient) :
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonersearchapi/api/PrisonerSearchApiApplicationClient.kt:7]: Needed because we are not using a request scope client on the jobs.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/PrisonRegime.kt (hunk 1 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/PrisonRegime.kt:33]: Not sure what the expiry time should actually be? Something configurable, and similar to NOMIS? Maybe a regime setting?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAllocationsService.kt [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      when (operation) {
+        DeallocateOperation.ENDING -> {
+          log.info("Ending allocations due to end today.")
+          allocationsDueToEnd().deallocate(now, "Allocation end date reached")
+        }
+
+        DeallocateOperation.EXPIRING -> {
+          log.info("Expiring allocations due to expire today.")
+          allocationsDueToExpire().deallocate(now, "Expired")
+        }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAllocationsService.kt:48]: Do we need to consider transactional level here - i.e. one transaction per deallocation, or one transaction covering all deallocations on the run of this job? Is the default to commit individually on the `saveAndFlush` of managed entities?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAllocationsService.kt [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.entity.PrisonRegime
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.entity.PrisonerStatus
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.repository.ActivityRepository
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.repository.ActivityScheduleRepository
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.repository.AllocationRepository
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.repository.PrisonRegimeRepository
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.repository.RolloutPrisonRepository
+import java.time.LocalDate
+import java.time.LocalDateTime
+
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAllocationsService.kt:22]: I like this separated into its own service.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReceivedEventHandler.kt (hunk 4 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReceivedEventHandler.kt:47]: No point in reconsidering as an interesting event.

**Review by timharrison-moj — APPROVED** [review]
A thought on transactions, but all looking good.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAllocationsService.kt (hunk 6 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAllocationsService.kt:48]: Possibly yes.  Lets review when we do the other job tickets you mention.
